### DISCUSSION TITLE
Rendering many autocompletes at once

### DISCUSSION THREAD

**Discussion by humandad** [floating-ui/floating-ui#2872] [Q&A] [answered]
I'm working on a project where I have a table (with many columns and many rows) where each cell contains an autocomplete input.

For each cell, I use the `useFloating`, `useListNavigation`, and `useInteractions` hooks. From some basic profiling, it seems like floating-ui is doing quite a lot of work when rendering these components. However, I really don't need any positioning work to be done at all when these components are rendered, I only need the positioning calculated when one of these autocompletes is activated by the `useListNavigation` hook.

Is there a way to defer the position calculations in floating-ui to only happen if the `open` property is set to true?

Thanks!

**Comment by atomiks** [comment, marked as answer]
As long as you're conditionally rendering them, not hiding them with CSS, the positioning work (and other stuff) is deferred.

**Reply by humandad** [reply]
Thanks!

**Comment by piecyk** [comment]
@atomiks I’ve seen a similar issue on our side. We have tooltips, popovers, and menus inside a table, and even with table virtualization the cost can still add up quickly when every cell initializes Floating UI hooks.

For example, in one rendered view we currently have around { tooltip: 242, popover: 64, menu: 6 } floating-related components mounted. Most of them are never opened during normal scrolling/rendering, so paying the setup/positioning cost upfront feels unnecessary.

Ideally, we’d like to avoid doing any positioning/listener work until the floating element is actually opened.

I was thinking about something like a lazy popover/tooltip wrapper. By default it would only render the reference element, and only on onMouseEnter, onFocus, or similar interaction would it mount the Floating UI logic/content.

**Reply by atomiks** [reply]
In Base UI I deferred some hooks until `open || mounted` https://github.com/mui/base-ui/pull/4661

**Reply by piecyk** [reply]
Thanks, that makes sense.

I was looking into the approach from mui/base-ui#4661, and it looks like the key idea is to split the reference-side interaction from the actual floating/positioning setup.

In our case, the main perf issue is that we can have hundreds of tooltip/popover triggers rendered in a table while scrolling, but most of them are never opened. So ideally we want the closed state to only set up a very small trigger layer, and avoid running useFloating() until the popup is actually opened.

I noticed useHover can work with useFloatingRootContext, so for tooltip/popover we could do something like:

1. create a root context with useFloatingRootContext
2. use useHover / useClick on the trigger side
3. keep only the reference element + open state mounted
4. mount the real floating content only when open === true

That seems close in spirit to what Base UI is doing with useHoverReferenceInteraction, but using the public Floating UI APIs
